packages/workers-editor-shared/lib/useInjectSources.tsx

```diff
@@ -0,0 +1,93 @@
+import { useCallback } from "react";
+import { Channel } from "./ipc";
+import type { FromErrorPage, SourcesLoadedMessage, ToErrorPage } from "./ipc";
+import type { TypedModule } from "./types";
+
+const encoder = new TextEncoder();
+const decoder = new TextDecoder();
+
+function getFilesFromModules(
+	entrypoint: string,
+	modules: Record<string, TypedModule>,
+	serviceWorkerScript: string | undefined
+) {
+	const isServiceWorker =
+		Object.entries(modules).length === 1 && serviceWorkerScript !== undefined;
+
+	const files = isServiceWorker
+		? [
+				{
+					path: entrypoint,
+					contents: encoder.encode(serviceWorkerScript),
+				},
+			]
+		: Object.entries(modules)
+				.filter(
+					([path, m]) =>
+						path.endsWith(".js") ||
+						m.type === "application/javascript" ||
+						m.type === "application/javascript+module"
+				)
+				.map(([path, m]) => ({
+					path,
+					contents: m.contents,
+				}));
+
+	let internalLines: number | undefined;
+	if (isServiceWorker && modules[entrypoint]) {
+		const totalScriptLength = serviceWorkerScript?.split("\n").length ?? 0;
+
+		const userScriptLength = decoder
+			.decode(modules[entrypoint].contents)
+			.split("\n").length;
+
+		internalLines = totalScriptLength - userScriptLength;
+	}
+
+	return { files, internalLines };
+}
+
+export function useInjectSources(
```

Comment:
Given that this is a fairly unique hook, it would be worth just describing what it does and what is its use.

As a kind of nit-addendum, I like to put the main export at the top of the file to make the code more readable. I don't know if this has been discussed elsewhere and a decision has been made, so this is just adding my two euro-cents to the conversation. When I open a file, I like to see immediately:

- what it imports,
- what it exports,
- and a description of what it does.

Reply:
I've added a description. With regards to the ordering, happy to change to a different order if you'd prefer. I generally tend to stick with exports at the bottom of a file, with all utility functions they depend on above them (so that you can read through a file and see how things are built up), but that's just a mild personal preference rather than anything else.

Reply:
Completely personal preference, I agree. Until someone brings out a paper on it I'm happy to go with whatever everyone else does.